src/installer/managed/Microsoft.NET.HostModel/ComHost/TypeLibReader.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+using System;
+using System.Buffers.Binary;
+
+namespace Microsoft.NET.HostModel.ComHost
+{
+    /// <summary>
+    /// Reads data from a COM Type Library file based on the official implementation.
+    /// </summary>
+    internal class TypeLibReader
+    {
+        private byte[] tlbBytes;
+
+        public TypeLibReader(byte[] tlbBytes)
+        {
+            this.tlbBytes = tlbBytes;
+        }
+
+        private const int OffsetOfGuidOffset = sizeof(int) * 2;
+        private const int SizeOfGuidOffset = sizeof(int);
+
+        private const int OffsetOfMajorVersion = OffsetOfGuidOffset + SizeOfGuidOffset + sizeof(int) * 2 + sizeof(ushort) * 2;
+        private const int SizeOfMajorVersion = sizeof(ushort);
+        private const int OffsetOfMinorVersion = OffsetOfMajorVersion + SizeOfMajorVersion;
+        private const int SizeOfMinorVersion = sizeof(ushort);
+
+        private const int OffsetOfTypeInfosCount = OffsetOfMinorVersion + SizeOfMinorVersion + sizeof(int);
+        private const int SizeOfTypeInfosCount = sizeof(int);
+
+        private const int OffsetOfTablesStart = OffsetOfTypeInfosCount + SizeOfTypeInfosCount + sizeof(int) * 12;
+        private const int NumTablesToSkip = 5;
+        private const int SizeOfTableHeader = sizeof(int) * 4;
+
+        private Guid FindGuid(ReadOnlySpan<byte> fileContents)
+        {
+            checked
+            {
+                int typelibGuidEntryOffset = (int)BinaryPrimitives.ReadUInt32LittleEndian(fileContents.Slice(OffsetOfGuidOffset));
+                int infoRefsOffsetCount = (int)BinaryPrimitives.ReadUInt32LittleEndian(fileContents.Slice(OffsetOfTypeInfosCount));
+                int infoBytes = infoRefsOffsetCount * SizeOfTypeInfosCount;
+                int guidTableOffset = OffsetOfTablesStart + infoBytes + SizeOfTableHeader * NumTablesToSkip;
+                int fileOffset = (int)BinaryPrimitives.ReadUInt32LittleEndian(fileContents.Slice(guidTableOffset));
+                return new Guid(fileContents.Slice(fileOffset + typelibGuidEntryOffset, 16).ToArray());
```

Comment:
I'm guessing we're doing all this reading ourselves (instead of using `LoadTypeLibEx`) so that it can run on non-Windows? Can you add a comment about why we do this?

Reply:
Actually, I guess not being able to embed clsidmap on non-Windows is always logged as an error, so I don't think cross-building works? Why do we need to do the reading ourselves?

Reply:
We're doing the reading ourselves so we don't need to worry about typelib dependencies being discoverable during the build.

Reply:
Can you add a comment (remarks on the class comment?) calling out that is why we need to do the reading ourselves?